PackageExplorer/Controls/MultiSelectTreeView.cs

```diff
@@ -0,0 +1,329 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Windows;
+using System.Windows.Controls;
+using System.Windows.Input;
+using System.Windows.Media;
+using System.Windows.Media.Media3D;
+
+namespace PackageExplorer.Controls
+{
+    public class MultiSelectTreeView : TreeView
+    {
+        public MultiSelectTreeView()
+        {
+            Loaded += (s, a) =>
+            {
+                GotFocus += OnTreeViewItemGotFocus;
+                PreviewMouseLeftButtonDown += OnTreeViewItemPreviewMouseDown;
+                PreviewMouseLeftButtonUp += OnTreeViewItemPreviewMouseUp;
+            };
+
+            Unloaded += (s, a) =>
+            {
+                GotFocus -= OnTreeViewItemGotFocus;
+                PreviewMouseLeftButtonDown -= OnTreeViewItemPreviewMouseDown;
+                PreviewMouseLeftButtonUp -= OnTreeViewItemPreviewMouseUp;
+            };
+        }
+
+        private static TreeViewItem? _selectTreeViewItemOnMouseUp;
+        private static DateTime _lastTime = DateTime.Now;
+
+        public static readonly DependencyProperty IsItemSelectedProperty = DependencyProperty.RegisterAttached("IsItemSelected", typeof(Boolean), typeof(MultiSelectTreeView), new PropertyMetadata(false, OnIsItemSelectedPropertyChanged));
+
+        public static bool GetIsItemSelected(TreeViewItem? element)
+        {
+            if (element != null)
+            {
+                return (bool)element.GetValue(IsItemSelectedProperty);
+            }
+
+            return false;
+        }
+
+        public static void SetIsItemSelected(TreeViewItem element, Boolean value)
+        {
+            if (element == null) return;
+
+            element.SetValue(IsItemSelectedProperty, value);
+        }
+
+        private static void OnIsItemSelectedPropertyChanged(DependencyObject d, DependencyPropertyChangedEventArgs e)
+        {
+            var treeViewItem = d as TreeViewItem;
+            var treeView = FindTreeView(treeViewItem);
+            if (treeViewItem != null && treeView != null)
+            {
+                var selectedItems = GetSelectedItems(treeView);
+                if (selectedItems != null)
+                {
+                    if (GetIsItemSelected(treeViewItem))
+                    {
+                        selectedItems.Add(treeViewItem.Header);
```

Comment:
I have debugged the `null` items and it looks like the problem is here.
Why are you using the `Header` instead of the `DataContext` here?
When I used the `DataContext` then no `null` items are added and everything works like before.

Reply:
I have switched Header to DataContext. But it was adding duplicates of treeViewItem.DataContext into SelectedItems list. So additionally I added check, whether it is not adding same item twice. :)